### ISSUE TITLE
nix eval crashes on any derivation: called with unexpected argument 'meta'

### ISSUE THREAD

**Issue by DavHau** [NixOS/nix#5447]
**Describe the bug**
When using nix eval on an expression which returns a derivation, nix crashes with:
`error: anonymous function at /nix/store/p53cz6rh27q40g9i0q98k3vfrz6lm12w-source/pkgs/build-support/fetchurl/boot.nix:5:1 called with unexpected argument 'meta', at /nix/store/p53cz6rh27q40g9i0q98k3vfrz6lm12w-source/pkgs/build-support/fetchpatch/default.nix:18:1`

This problem occurs with nix 2.3 and 2.4.

**Steps To Reproduce**
execute:
```
nix eval '(import <nixpkgs> {}).hello'
```

**Expected behavior**

print out the derivation path like the repl does

**`nix-env --version` output**
nix (Nix) 2.3.16

**Additional context**
Some might argue that `nix eval` should not be used to evaluate derivations in the first place, but if this is the case, then nix should error out with a message saying that. Right now the stack trace led me to debug my code for hours until I noticed that it's just nix not being able to evaluate that. (Not complaining, just underlining the importance of that issue being fixed.)

**Comment by mikepurvis** [comment]
I'm seeing this too; I'm surprised the final 2.4 version shipped without this corrected.

**Comment by regnat** [comment]
That might just be a purely theoretical argument, but it’s technically not forbidden to run `nix eval` on a derivation, the error is just a detail of how nixpkgs works:

```console
$ nix eval --expr 'builtins.derivation {name = "foo"; builder = "/bin/sh"; system = "x86_64-linux"; }'
{ all = [ { all = <CYCLE>; builder = "/bin/sh"; drvAttrs = { builder = "/bin/sh"; name = "foo"; system = "x86_64-linux"; }; drvPath = "/nix/store/9d794zbx10n43fa6zaqf33pfymdrm4r2-foo.drv"; name = "foo"; out = { all = <CYCLE>; builder = "/bin/sh"; drvAttrs = { builder = "/bin/sh"; name = "foo"; system = "x86_64-linux"; }; drvPath = "/nix/store/9d794zbx10n43fa6zaqf33pfymdrm4r2-foo.drv"; name = "foo"; out = <CYCLE>; outPath = "/nix/store/xpcvxsx5sw4rbq666blz6sxqlmsqphmr-foo"; outputName = "out"; system = "x86_64-linux"; type = "derivation"; }; outPath = "/nix/store/xpcvxsx5sw4rbq666blz6sxqlmsqphmr-foo"; outputName = "out"; system = "x86_64-linux"; type = "derivation"; } ]; builder = "/bin/sh"; drvAttrs = { builder = "/bin/sh"; name = "foo"; system = "x86_64-linux"; }; drvPath = "/nix/store/9d794zbx10n43fa6zaqf33pfymdrm4r2-foo.drv"; name = "foo"; out = { all = [ { all = <CYCLE>; builder = "/bin/sh"; drvAttrs = { builder = "/bin/sh"; name = "foo"; system = "x86_64-linux"; }; drvPath = "/nix/store/9d794zbx10n43fa6zaqf33pfymdrm4r2-foo.drv"; name = "foo"; out = <CYCLE>; outPath = "/nix/store/xpcvxsx5sw4rbq666blz6sxqlmsqphmr-foo"; outputName = "out"; system = "x86_64-linux"; type = "derivation"; } ]; builder = "/bin/sh"; drvAttrs = { builder = "/bin/sh"; name = "foo"; system = "x86_64-linux"; }; drvPath = "/nix/store/9d794zbx10n43fa6zaqf33pfymdrm4r2-foo.drv"; name = "foo"; out = <CYCLE>; outPath = "/nix/store/xpcvxsx5sw4rbq666blz6sxqlmsqphmr-foo"; outputName = "out"; system = "x86_64-linux"; type = "derivation"; }; outPath = "/nix/store/xpcvxsx5sw4rbq666blz6sxqlmsqphmr-foo"; outputName = "out"; system = "x86_64-linux"; type = "derivation"; }
```

**Comment by balsoft** [comment]
I don't think this is a bug in nix. It's just the way nixpkgs works (basically you can't deeply evaluate the derivations there). If you need to just get the path, try `--raw`, which will stringify the expression result before printing it out:

```
$ nix eval nixpkgs#hello --raw
/nix/store/15c875mwri8xx3s0gqsdkdw7sqqyv55c-hello-2.10
```

**Comment by balodja** [comment]
As a workaround `toString` builtin may be used to explicitly convert derivation to string:
```
nix eval '(toString (import <nixpkgs> {}).hello)' --raw
```

Note that that won't be built without explicit instruction.

**Comment by DavHau** [comment]
Thanks for all your comments which helped a lot understanding the issue.

The fact that this is technically more an issue of nixpkgs than nix itself is irrelevant for the user I'd argue.
Practically nix isn't any useful without nixpkgs and vice versa. Users should be able to expect proper integration of the two and therefore an error message being displayed would be appropriate.

I see two possible options:
  1. Nixpkgs derivations are identified by having an attribute `type = "derivation"`. If this can be considered an ecosystem wide standard, then `nix eval` should never try to deeply evaluate attribute sets of that type and instead print an error or the string representation of the derivation (store path etc.).
  1. This should be fixed in nixpkgs somehow, so that an appropriate error is raised when something is evaluated that shouldn't be.

**Comment by erikarvstedt** [comment]
Duplicate of https://github.com/NixOS/nix/issues/5731, which has further discussions on how to fix it.
